File: sdk/cpp/src/api/trustchain.cpp
// Copyright 2026 Hypermesh Foundation. All rights reserved.
// Licensed under the Business Source License 1.1.

#include <hypermesh/api/trustchain.hpp>

namespace hypermesh {

TrustChainCertificateList TrustChainApi::certificates() const {
    auto j = http_.get("/api/v1/trustchain/certificates");
    return j.get<TrustChainCertificateList>();
}

TrustChainCertificate TrustChainApi::issue(const std::string& subject,
                                            const std::string& scope) const {
    auto body = nlohmann::json{{"subject", subject}, {"scope", scope}};
    auto j = http_.post("/api/v1/trustchain/issue", body);
    return j.get<TrustChainCertificate>();
}

TrustChainValidationResult TrustChainApi::validate(const std::string& cert_pem) const {
    auto body = nlohmann::json{{"cert_pem", cert_pem}};
    auto j = http_.post("/api/v1/trustchain/validate", body);
    return j.get<TrustChainValidationResult>();
}

TrustChainRevokeResult TrustChainApi::revoke(const std::string& cert_id) const {
    auto body = nlohmann::json{{"cert_id", cert_id}};
    auto j = http_.post("/api/v1/trustchain/revoke", body);
    return j.get<TrustChainRevokeResult>();
}

TrustChainDnsZoneList TrustChainApi::dns_zones() const {
    auto j = http_.get("/api/v1/trustchain/dns/zones");
    return j.get<TrustChainDnsZoneList>();
}

} // namespace hypermesh

File: sdk/typescript/src/api/trustchain.ts
import type { HttpClient } from "../client.js";
import type {
  TrustChainCertificate,
  TrustChainCertificateList,
  TrustChainDnsZoneList,
  TrustChainRevokeResult,
  TrustChainValidationResult,
} from "../types.js";

export class TrustChainApi {
  constructor(private readonly http: HttpClient) {}

  async certificates(): Promise<TrustChainCertificateList> {
    return this.http.get<TrustChainCertificateList>(
      "/api/v1/trustchain/certificates",
    );
  }

  async issue(
    subject: string,
    scope: string,
  ): Promise<TrustChainCertificate> {
    return this.http.post<TrustChainCertificate>(
      "/api/v1/trustchain/issue",
      { subject, scope },
    );
  }

  async validate(certPem: string): Promise<TrustChainValidationResult> {
    return this.http.post<TrustChainValidationResult>(
      "/api/v1/trustchain/validate",
      { cert_pem: certPem },
    );
  }

  async revoke(certId: string): Promise<TrustChainRevokeResult> {
    return this.http.post<TrustChainRevokeResult>(
      "/api/v1/trustchain/revoke",
      { cert_id: certId },
    );
  }

  async dnsZones(): Promise<TrustChainDnsZoneList> {
    return this.http.get<TrustChainDnsZoneList>(
      "/api/v1/trustchain/dns/zones",
    );
  }
}
